src/base/utils/tar.cpp

```diff
@@ -0,0 +1,134 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "tar.h"
+
+#include <cstring>
+
+#include <QByteArray>
+#include <QHash>
+#include <QString>
+
+namespace
+{
+    const int TAR_HEADER_SIZE = 512;
+
+    bool tarEOF(const char *data)
+    {
+        // static variables are auto initialized to zero
+        static char testBlock[2 * TAR_HEADER_SIZE];
+        return std::memcmp(testBlock, data, 2 * TAR_HEADER_SIZE) == 0;
+    }
+
+    class TarFile
+    {
+    public:
+        TarFile(const char *data)
+            : m_data(data)
+        {}
+
+        QString name() const
+        {
+            if (m_data && (*(m_data + 99) != '\0'))
+                return {};
+
+            return QString{m_data};
+        }
+
+        uchar type() const
+        {
+            if (!m_data) return 255; // zero denotes regular file
+
+            const char *type = m_data + 156;
+            return type ? (*type - '0') : 0;
+        }
+
+        uint size() const
+        {
+            if (!m_data) return 0;
+
+            return octalToDec(m_data + 124, 11);
+        }
+
+        QByteArray fileData() const
+        {
+            if (!m_data) return {};
+
+            return QByteArray{m_data + TAR_HEADER_SIZE, static_cast<int>(size())};
+        }
+
+    private:
+        uint octalToDec(const char *data, uint size) const
+        {
+            if (!data) return 0;
+
+            uint fileSize = 0;
+            for (uint i = 0; i < size; ++i)
+                fileSize = (8 * fileSize) + (data[i] - '0');
+
+            return fileSize;
+        }
+
+        const char *m_data;
+    };
+
+}
+
+
+QHash<QString, QByteArray> Utils::Tar::decompress(const QByteArray &data, bool *ok)
+{
+    // The EOF is **at least** 2 zero-ed headers. So valid data are 1 header + payload + 2 zero-ed headers
+    if (data.size() < 3 * TAR_HEADER_SIZE + 1) {
+        if (ok)
+            *ok = false;
```

Comment:
would be simpler if you just assigned `*ok = false;` at the top of this function.